### PR TITLE
[RW-5679][risk=no] Support Institution Table

### PR DESCRIPTION
**Pull request by jaycarlton**
Institution table upload supported. Added support to user table for denormalized columns from the affiliation table.

Fixed missing cohort upload and added institution support to DML (insert query) implementation for upload. Made some small timing logging and time printing improvements.

TODO:
* fix refex pattern/match for getting table name from an insert query. 

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI + API server with `yarn test-local`
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/test/java/org/pmiops/workbench/db/dao/InstitutionDaoTest.java [hunk 1]
```diff
@@ -106,4 +112,55 @@ public void test_uniqueDisplayNameRequired() {
...
+    assertThat(institutions.get(0).getDuaTypeEnum()).isEqualTo(DuaType.RESTRICTED);
+    assertThat(institutions.get(0).getOrganizationTypeEnum()).isEqualTo(OrganizationType.OTHER);
+    assertThat(institutions.get(0).getOrganizationTypeOtherText()).isEqualTo("Pyramid Scheme");
+
+    assertThat(institutions.get(1).getDisplayName()).isEqualTo("MASH");
+    assertThat(institutions.get(1).getDuaTypeEnum()).isEqualTo(DuaType.MASTER);
+    assertThat(institutions.get(1).getOrganizationTypeEnum())
+        .isEqualTo(OrganizationType.HEALTH_CENTER_NON_PROFIT);
+    assertThat(institutions.get(1).getDisplayName()).isEqualTo("MASH");
+  }
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/db/dao/InstitutionDaoTest.java:165]: does `getOrganizationTypeOtherText()` return NULL or `''` for inst 2?  Add an assert for the correct one

File: api/src/test/java/org/pmiops/workbench/reporting/ReportingMapperTest.java [hunk 2]
```diff
@@ -26,16 +30,33 @@
   public static class conifg {}
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/reporting/ReportingMapperTest.java:30]: ?

File: api/src/test/java/org/pmiops/workbench/testconfig/FakeTicker.java [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
 
 public class FakeTicker extends Ticker {
-  private final long elapsedMillis;
+  private long elapsedMillis;
 
-  protected FakeTicker(long elapsedMillis) {
+  protected FakeTicker(long tickNanos) {
     super();
-    this.elapsedMillis = elapsedMillis;
+    this.elapsedMillis = tickNanos;
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/testconfig/FakeTicker.java:11]: no

File: api/src/test/java/org/pmiops/workbench/testconfig/FakeTicker.java [hunk 4]
```diff
@@ -1,17 +1,19 @@
...
   }
 
   @Override
   public long read() {
+    elapsedMillis += TimeUnit.MILLISECONDS.toNanos(200);
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/testconfig/FakeTicker.java:16]: also no

**Review by jmthibault79 — CHANGES_REQUESTED** [review]
Mostly looks good.  A few nits and a few required fixes.

File: api/reporting/schemas/reporting-wizard.rb [hunk 5]
```diff
@@ -172,6 +179,29 @@ def to_output_path(subdir_name, suffix)
...
+             :bigquery => 'STRING',
+             :default_constant_value => 'OrganizationType.ACADEMIC_RESEARCH_INSTITUTION'
+         }
+     },
+     'user_verified_institutional_affiliation' => {
+         'institutional_role_enum' => {
+             :projection => 'InstitutionalRole',
+             :swagger => 'InstitutionalRole',
+             :bigquery => 'STRING',
+             :default_constant_value => 'InstitutionalRole.UNDERGRADUATE'
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply api/reporting/schemas/reporting-wizard.rb:203]: These only apply to unit test object instantiation. Only thing we care about is that it's a valid value.

File: api/src/main/java/org/pmiops/workbench/reporting/ReportingMapper.java [hunk 6]
```diff
@@ -17,16 +20,68 @@
...
+        .captureTimestamp(snapshotTimestamp)
+        .cohorts(toModelList(queryResultBundle.getCohorts()))
+        .institutions(toReportingInstitutionList(queryResultBundle.getInstitutions()))
+        .users(toReportingUserList(queryResultBundle.getUsers()))
+        .workspaces(toReportingWorkspaceList(queryResultBundle.getWorkspaces()));
+  }
+
+  // Define immutable value class to hold results of queries within a transaction. Mapping to
+  // Reporting DTO classes will happen outside the transaction.
+  class QueryResultBundle {
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/reporting/ReportingMapper.java:54]: It's an intermediate object that only exists to make timings easier. The mapper has the only methods that accept it, so it made sense to me to put it there. If it's in the service, then it has to be public, at which point it should be its own class. So I think this is the best encapsulation.

File: api/src/main/java/org/pmiops/workbench/reporting/ReportingMapper.java [hunk 7]
```diff
@@ -17,16 +20,68 @@
     config = MapStructConfig.class,
     uses = {CommonMappers.class, DbStorageEnums.class})
 public interface ReportingMapper {
-  ReportingUser toDto(ProjectedReportingUser prjUser);
+
+  ReportingInstitution toReporrtingInstitution(ProjectedReportingInstitution prjInstitution);
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply api/src/main/java/org/pmiops/workbench/reporting/ReportingMapper.java:24]: My laptop keyboard is over-correcting for not being able to type keys by auto-repeating (even with the key repeat feature disabled). If I weren't an Alphabet shareholder I'd requisition a new one.

File: api/src/test/java/org/pmiops/workbench/db/dao/InstitutionDaoTest.java (hunk 1 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/db/dao/InstitutionDaoTest.java:165]: Projections don't do any conversion at all to what's in the DB, so you get whatever is in the table. If the application enforces nullity, that's great, but I don't think we need tests to show that all nullable columns can be null.  

If we're going to tackle null vs empty, I'd prefer to have a standalone story for that to do it app-wide, maybe even in an interceptor 😈 .

File: api/src/test/java/org/pmiops/workbench/reporting/ReportingMapperTest.java (hunk 2 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/reporting/ReportingMapperTest.java:30]: renamed. These classes are basically never used by name (since annotation finds them), but `Configuration` is our standard name.

File: api/src/test/java/org/pmiops/workbench/testconfig/FakeTicker.java (hunk 3 above)
Comment (jaycarlton) [reply api/src/test/java/org/pmiops/workbench/testconfig/FakeTicker.java:11]: If I had a dollar for every timestamp and duration type I've had to use in this project, I'd have $150 Eastern/$149 Central.

File: api/src/main/resources/workbench-api.yaml [hunk 8]
```diff
@@ -8446,9 +8467,46 @@ definitions:
       name:
         type: string
         description: User-provided human-readable name for this cohort.
+      type:
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply api/src/main/resources/workbench-api.yaml:8470]: Another good argument for moving these non-API classes into their own Swagger file and output directory.

@aweng98 Here's a good example of what we were talking about the other day: a Swagger change the UI but doesn't force the UI unit tests to run.
```
No relevant changes in ui directory on non-master branch. halting job.
```

File: api/src/test/java/org/pmiops/workbench/db/dao/InstitutionDaoTest.java (hunk 1 above)
Comment (jmthibault79) [reply api/src/test/java/org/pmiops/workbench/db/dao/InstitutionDaoTest.java:165]: ok

**Comment by jaycarlton** [comment]
I did some rebasing so I could rebase the terra branch against this one. Hope that didn't break everything.

**Review by jmthibault79 — APPROVED** [review]
